mscore/instrdialog.cpp

```diff
@@ -409,10 +410,62 @@ void MuseScore::editInstrList()
       //
 
       int n = rootScore->nstaves();
+      int curSpan = 0;
       for (int i = 0; i < n; ++i) {
             Staff* staff = rootScore->staff(i);
-            if (staff->barLineSpan() > (n - i))
-                  rootScore->undoChangeBarLineSpan(staff, n - i, 0, (rootScore->staff(n-1)->lines()-1) * 2);
+            int span = staff->barLineSpan();
+            int setSpan = -1;
+
+            // determine if we need to update barline span
+            if (curSpan == 0) {
+                  // no current span; this staff must start a new one
+                  if (span == 0) {
+                        // no span; this staff must have been within a span
+                        // update it to a span of 1
+                        setSpan = 1;
+                        }
+                  else if (span > (n - i)) {
+                        // span too big; staves must have been removed
+                        // reduce span to last staff
+                        setSpan = n - i;
+                        }
+                  else if (span > 1 && staff->barLineTo() > 0) {
+                        // TODO: check if span is still valid
```

Comment:
So, here is the regression:

If you have a barline span of more than staff and it ends somewhere in the middle of the last staff, this customization will be lost if you make changes in Edit / Instruments.  The barline will be reset to end at the bottom of the staff as it normally would.  Workround: you can then restore the customized span and it will work fine.

I judge this to be not worth worrying about.  Barlines spans of a *single* staff that end in the middle of a staff are common enough (eg, "tick" barlines), as are barlines spans of multiple staves that end *above* the last staff ("Mensurstrich").  I was able to make these cases work.  I don't know of a use case for barline spans of *multiple* staves that end in the *middle* of a staff.  I'm sure there is one, but if that is the only thing that gets messed up on Edit / Instruments, I would gladly accept that for now in order to fix the more common & more serious problems that this PR addresses.